### PR TITLE
[p4fmt]: print comments

### PR DESCRIPTION
**Pull request by snapdgn**
This PR is an initial attempt to print comments using p4fmt.It is not intended to be merged (atleast for now); it serves as a preliminary draft to initiate discussion and gather feedback from the community on potential approaches moving forward.

 PR #1226 parsed and saved comments as part of the `InputSources` data structure for further use.

This PR outlines an approach where we take in that position information and associate it with the nodes _while_ printing with [`toP4` ](https://github.com/p4lang/p4c/blob/main/frontends/p4/toP4/toP4.h). For this to work we need to modify _each_ node type of `toP4` to access comments related to that particular node(based solely on position info) and print accordingly. 
(The PR demonstrates this wherein comments are printed related to `Typedef` nodes, It only associates the comments which are closest, things like comments on same line are not handled, multi-line comments are not handled, efficiency is not taken into account etc. This is just an intial attempt and code is very _hack-ish_).

*Sample Input*:
```
typedef bit<9>  egressSpec_t;
// comm1
typedef bit<48> macAddr_t;
// comm2
typedef bit<32> ip4Addr_t;

header ethernet_t {
    macAddr_t dstAddr;
    // Comm3
    macAddr_t srcAddr;
    bit<16>   etherType;
}
```
*Sample Output*:
```
typedef bit<9> egressSpec_t;
// comm1
typedef bit<48> macAddr_t;
// comm2
typedef bit<32> ip4Addr_t;
header ethernet_t {
    macAddr_t dstAddr;
    macAddr_t srcAddr;
    bit<16>   etherType;
}
```
A more sophisticated [approach](https://jayconrod.com/posts/129/preserving-comments-when-parsing-and-formatting-code) would be to include a separate common comments field in each node. By using the comment position information along with the nodes, we could traverse the AST nodes and associate the comments with the relevant nodes.

### CODE REVIEW

File: frontends/p4/toP4/toP4.cpp [hunk 1]
```diff
@@ -241,6 +241,9 @@ bool ToP4::preorder(const IR::Type_Void *) {
 }
 
 bool ToP4::preorder(const IR::Type_Name *t) {
+    auto cmts = t->srcInfo.getComments();
+    builder.append(cmts);
```
Comment (fruffy) [inline frontends/p4/toP4/toP4.cpp:245]: You are seeing these failures because this always prints a newline unconditionally. Would suggest only adding it when there are comments.

File: lib/source_file.h [hunk 2]
```diff
@@ -267,6 +273,7 @@ class InputSources final {
 
  public:
     InputSources();
+    friend class SourceInfo;
```
Comment (fruffy) [inline lib/source_file.h:276]: Why is this needed? Maybe we should add the appropriate public getter setters.

**Comment by fruffy** [comment]
I would also recommend adding an option toP4 to toggle whether to add comments etc. This way you do not conflict with the current reference files in the compiler.

File: frontends/p4/toP4/toP4.cpp [hunk 3]
```diff
@@ -281,6 +284,9 @@ bool ToP4::preorder(const IR::Type_Typedef *t) {
         visit(t->annotations);
         builder.spc();
     }
+    auto cmts = t->srcInfo.getComments();
+    builder.append(cmts);
+    builder.append("\n");
```
Comment (qobilidop) [inline frontends/p4/toP4/toP4.cpp:289]: I guess comments before annotations might make more sense here? For example:

```p4
\\ comments
@anno
typedef a b;
```

**Comment by snapdgn** [comment]
Thanks for the review :), I'll resolve those.
Also, I'll go ahead with a separate transformative pass as discussed in the meeting for attaching nodes with comments, so that we can gradually build up on it & merge it upstream in smaller chunks.

**Comment by qobilidop** [comment]
@snapdgn Thanks for sharing the blog post [Preserving comments when parsing and formatting code](https://jayconrod.com/posts/129/preserving-comments-when-parsing-and-formatting-code), I find it pretty informative. Especially the description on the algorithm for attaching comments to AST nodes, has made it much clearer to me how this could be done.

File: lib/source_file.h [hunk 4]
```diff
@@ -230,13 +232,17 @@ struct SourceFileLine {
...
-    SourceInfo srcInfo;
+    SourcePosition srcPos;
     bool singleLine;
     cstring body;
 
  public:
-    Comment(SourceInfo srcInfo, bool singleLine, cstring body)
-        : srcInfo(srcInfo), singleLine(singleLine), body(body) {}
+    Comment(SourcePosition srcPos, bool singleLine, cstring body)
+        : srcPos(srcPos), singleLine(singleLine), body(body) {}
```
Comment (asl) [inline lib/source_file.h:241]:
```suggestion
        : srcPos(std::move(srcPos)), singleLine(singleLine), body(body) {}
```

File: lib/source_file.cpp [hunk 5]
```diff
@@ -69,7 +71,35 @@ void InputSources::addComment(SourceInfo srcInfo, bool singleLine, cstring body)
     if (!singleLine)
         // Drop the "*/"
         body = body.exceptLast(2);
-    comments.push_back(new Comment(srcInfo, singleLine, body));
+    auto srcPos = SourcePosition(srcInfo.getStart());
+
+    comments.push_back(new Comment(srcPos, singleLine, body));
```
Comment (asl) [inline lib/source_file.cpp:76]: construct `srcPos` in place (as a temporary) or move in place to take advantage of moving.

File: lib/source_file.cpp [hunk 6]
```diff
@@ -69,7 +71,35 @@ void InputSources::addComment(SourceInfo srcInfo, bool singleLine, cstring body)
...
+}
+
+std::string SourceInfo::getComments() const {
+    if ((sources == nullptr)) {
```
Comment (asl) [inline lib/source_file.cpp:80]:
```suggestion
    if (sources == nullptr) {
```

File: lib/source_file.cpp [hunk 7]
```diff
@@ -69,7 +71,35 @@ void InputSources::addComment(SourceInfo srcInfo, bool singleLine, cstring body)
...
+        return "";
+    }
+
+    std::ostringstream oss;
+    uint targetLine = start.getLineNumber();
```
Comment (asl) [inline lib/source_file.cpp:85]: This returns `unsigned`, not `uint`. There is no `uint` standard type.
```suggestion
    unsigned targetLine = start.getLineNumber();
```

File: lib/source_file.cpp [hunk 8]
```diff
@@ -69,7 +71,35 @@ void InputSources::addComment(SourceInfo srcInfo, bool singleLine, cstring body)
...
+    const Comment *closestComment = nullptr;
+
+    for (const auto &comment : sources->comments) {
+        uint commentLine = comment->getSourcePosition().getLineNumber();
```
Comment (asl) [inline lib/source_file.cpp:89]: ditto

File: lib/source_file.cpp [hunk 9]
```diff
@@ -69,7 +71,35 @@ void InputSources::addComment(SourceInfo srcInfo, bool singleLine, cstring body)
...
+        if (commentLine < targetLine &&
+            (closestComment == nullptr ||
+             commentLine > closestComment->getSourcePosition().getLineNumber()) &&
+            (diff == 0 || diff == 1)) {
+            closestComment = comment;
+        }
+    }
+
+    if (closestComment != nullptr) {
+        oss << closestComment->toString();
```
Comment (asl) [inline lib/source_file.cpp:100]: What is the point of this terrible code? You're using `std::ostringstream` just to do conversion from `cstring` to `std::string`? This at least involves multiple string copies: first into the stream, and then back. Consider rewriting the code in more straightforward manner, e.g.:

```c++
return closestComment ? closestComment->toString().string() : "";
```

File: lib/source_file.cpp [hunk 10]
```diff
@@ -69,7 +71,35 @@ void InputSources::addComment(SourceInfo srcInfo, bool singleLine, cstring body)
     if (!singleLine)
         // Drop the "*/"
         body = body.exceptLast(2);
-    comments.push_back(new Comment(srcInfo, singleLine, body));
+    auto srcPos = SourcePosition(srcInfo.getStart());
+
+    comments.push_back(new Comment(srcPos, singleLine, body));
+}
+
+std::string SourceInfo::getComments() const {
```
Comment (asl) [inline lib/source_file.cpp:79]: The name is misleading. The function returns a single comment, why "comments" then?

**Review by asl — CHANGES_REQUESTED** [review]

**Comment by snapdgn** [comment]
Thanks for all the reviews and comments, they've been incredibly helpful. :)
I am convinced that this requires a major overhaul since the process of attaching comments needs to be addressed in a separate AST pass.
I'll ensure to incorporate all the suggested changes in the rework.
